grpc/http/forwarders.go

```diff
@@ -133,6 +138,53 @@ type StreamForwarderFunc func(
 	opts ...func(context.Context, http.ResponseWriter, proto.Message) error,
 )
 
+func writeKeepalive(w http.ResponseWriter, mut *sync.Mutex) {
+	mut.Lock()
+	defer mut.Unlock()
+
+	// Per https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation,
+	// lines that start with a `:` must be ignored by the client.
+	_, err := w.Write([]byte(":\n"))
+
+	if err != nil {
+		log.Warnf("failed to write http keepalive response: %v", err)
+	} else if f, ok := w.(http.Flusher); ok {
+		f.Flush()
+	}
+}
+
+func keepalive(ctx context.Context, w http.ResponseWriter, mut *sync.Mutex, t time.Ticker) {
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			writeKeepalive(w, mut)
+		}
+	}
+}
+
+func withKeepalive(ctx context.Context, w http.ResponseWriter) http.ResponseWriter {
+	keepaliveInterval := time.Duration(time.Second * 15)
+	keepaliveTicker := time.NewTicker(keepaliveInterval)
+	mut := sync.Mutex{}
+
+	go keepalive(ctx, w, &mut, *keepaliveTicker)
+
+	return httpsnoop.Wrap(w, httpsnoop.Hooks{
+		Write: func(next httpsnoop.WriteFunc) httpsnoop.WriteFunc {
+			return func(p []byte) (int, error) {
+				mut.Lock()
+				defer mut.Unlock()
+				keepaliveTicker.Reset(keepaliveInterval)
```

Comment:
I still think you should move the ticker into keepalive and don't bother with the `Reset` - it can't see it needed

Reply:
I can do that.  Pretty much the only downside is additional keepalive messages being sent periodically - not a huge deal.

Reply:
I'm aware that I'm pushing you for the most minimal solution for this, ever asking you to remove tests. I think having less to maintain is for the best. I think a lot of users will be really happy with this change.